File: QuickStarts/VB/Stoplight/Src/Logic/Stoplight.vb
'===============================================================================
' Microsoft patterns & practices
' Unity Application Block
'===============================================================================
' Copyright © Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================

Imports System
Imports Microsoft.Practices.Unity
Imports StopLight.ServiceImplementations
Imports StopLight.ServiceInterfaces

Namespace StopLight.Logic
    Public Enum StoplightColors
        Green
        Yellow
        Red
    End Enum

    Public Class LightChangedEventArgs
        Inherits EventArgs
        Private currColor As StoplightColors

        Public Property CurrentColor() As StoplightColors
            Get
                Return currColor
            End Get
            Private Set(ByVal value As StoplightColors)
                currColor = Value
            End Set
        End Property

        Public Sub New(ByVal color As StoplightColors)
            currColor = color
        End Sub
    End Class

    Public Delegate Sub StoplightChangedHandler(ByVal sender As Object, ByVal e As LightChangedEventArgs)

    Public Class Stoplight
        Public Const Green As StoplightColors = StoplightColors.Green
        Public Const Yellow As StoplightColors = StoplightColors.Yellow
        Public Const Red As StoplightColors = StoplightColors.Red

        Private currColor As StoplightColors = StoplightColors.Green
        Private _logger As ILogger = New NullLogger()
        Public Event Changed As StoplightChangedHandler

        Public ReadOnly Property CurrentColor() As StoplightColors
            Get
                Return currColor
            End Get
        End Property

        <Dependency()> _
        Public Property Logger() As ILogger
            Get
                Return _logger
            End Get
            Set(ByVal value As ILogger)
                _logger = value
            End Set
        End Property

        Public Sub [Next]()
            currColor += 1
            If currColor > StoplightColors.Red Then
                currColor = StoplightColors.Green
            End If
            RaiseChanged()
            _logger.Write(String.Format("LIGHT CHANGED TO {0}", CurrentColor))
        End Sub

        Protected Sub RaiseChanged()
            Dim handlers As StoplightChangedHandler = ChangedEvent
            If Not handlers Is Nothing Then
                Dim e As New LightChangedEventArgs(currColor)
                handlers(Me, e)
            End If
        End Sub

    End Class
End Namespace

File: quickstarts/CS/Stoplight/Src/Logic/Stoplight.cs
// Copyright (c) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.

using System;
using Microsoft.Practices.Unity;
using StopLight.ServiceImplementations;
using StopLight.ServiceInterfaces;

namespace StopLight.Logic
{
	public enum StoplightColors
	{
		Green,
		Yellow,
		Red
	}

	public class LightChangedEventArgs : EventArgs
	{
		private StoplightColors currentColor;

		public StoplightColors CurrentColor
		{
			get { return currentColor; }
			private set { currentColor = value; }
		}

		public LightChangedEventArgs(StoplightColors color)
		{
			CurrentColor = color;
		}
	}

	public delegate void StoplightChangedHandler(object sender, LightChangedEventArgs e);

	public class Stoplight
	{
		public const StoplightColors Green = StoplightColors.Green;
		public const StoplightColors Yellow = StoplightColors.Yellow;
		public const StoplightColors Red = StoplightColors.Red;

		private StoplightColors currentColor = StoplightColors.Green;
		private ILogger logger = new NullLogger();
		public event StoplightChangedHandler Changed;

		public StoplightColors CurrentColor
		{
			get { return currentColor; }
		}

		[Dependency]
		public ILogger Logger
		{
			get { return logger; }
			set { logger = value; }
		}

		public void Next()
		{
			++currentColor;
			if(currentColor > StoplightColors.Red)
			{
				currentColor = StoplightColors.Green;
			}
			RaiseChanged();
			logger.Write(string.Format("LIGHT CHANGED TO {0}", currentColor));
		}

		protected void RaiseChanged()
		{
			StoplightChangedHandler handlers = Changed;
			if(handlers != null)
			{
				LightChangedEventArgs e = new LightChangedEventArgs(currentColor);
				handlers(this, e);
			}
		}
	}
}
